frontend/src/features/public-form/PublicFormProvider.tsx

```diff
@@ -388,6 +389,33 @@ export const PublicFormProvider = ({
     /* enabled= */ !submissionData,
   )
 
+  const {
+    data: encryptedPreviousSubmission,
+    isLoading: isSubmissionLoading,
+    error: encryptedSubmissionError,
+  } = useEncryptedSubmission(
+    formId,
+    previousSubmissionId,
+    // Stop querying once submissionData is present.
+    /* enabled= */ !submissionData,
+  )
+
+  // Replace form fields, logic, and workflow with the previous snapshotted form definition for MRF >= 2nd step consistency.
+  // This is safe to use for storage mode and 1st step of MRF which uses the latest form definition,
+  // while MRF >= 2nd step uses the snapshotted form definition.
+  const mrfConsistentFormData =
+    latestNonMrfConsistentFormData && encryptedPreviousSubmission
+      ? {
+          ...latestNonMrfConsistentFormData,
+          form: {
+            ...latestNonMrfConsistentFormData.form,
+            form_fields: encryptedPreviousSubmission.form_fields,
+            form_logics: encryptedPreviousSubmission.form_logics,
+            workflow: encryptedPreviousSubmission.workflow,
+          },
+        }
+      : latestNonMrfConsistentFormData
+
```

Comment:
1. this might need a useMemo hook!
2. what do you think about calling this `data`, since this is the only use of `latestNonMrfConsistentFormData`?
3. good idea to put the annotation (the fact that this is computed from the submission snapshot when steps >= 2) in the comments: it'll show up in an IDE's context window.

<img width="962" height="631" alt="image" src="https://github.com/user-attachments/assets/1a2a6f22-7f2f-46a0-a07d-28eaa9090c9a" />